Webui autopart v1

rvykydal opened the pull request:
First iteration of guided partitioning (autopartitioning) in WebUI.

There is a number of issues that need to be done either in scope of this PR or (preferably) as a follow-up tasks.
In this PR I would like to focus on doing the things correctly code-wise. 

Outstanding issues:

Must address in this PR:
- [x] Fix existing tests.
- [x] Add tests for the new functionality.
- [x] Test that the installation really works.
- [x] Add `idPrefix` to some ids.
- [x] Address unit tests failures related to strings.
- [x] Fix the devices(s) typos

May address here or separately:
- [ ] Change "Checking disks" to "Checking storage configuration" (applyConfiguration on Next after setting password) Related to [INSTALLER-3419](https://issues.redhat.com/browse/INSTALLER-3419)
- [ ] [INSTALLER-3447](https://issues.redhat.com/browse/INSTALLER-3447) Fix tooltip on hover over partitionong option radio does not work on the preview image. Consider also removing the no-entry symbol.
- [ ]  [INSTALLER-3443](https://issues.redhat.com/browse/INSTALLER-3443) Fix spacing of the alert info for disabled choices on Storage Configuration screen.
- [ ] [INSTALLER-3444](https://issues.redhat.com/browse/INSTALLER-3444) Fix vertical centering of the Storage Configuration choices.
- [ ] [INSTALLER-3445](https://issues.redhat.com/browse/INSTALLER-3445) Implement "Review and install" screen according to the mockup (show selections).
- [x] Update autopart option details / drawer content (there are just placeholders / raw phrases on some places)
- [ ] [INSTALLER-3405](https://issues.redhat.com/browse/INSTALLER-3405) Fix help drawer to be closed on leaving a screen (to prevent https://rvykydal.fedorapeople.org/webui/autopart/autopart_partitioning_choice_drawer.webm) 
- [ ] [INSTALLER-3446](https://issues.redhat.com/browse/INSTALLER-3446) Fix the state of disk selection screen after applying partitioning and going back (https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screencasts/pr_autopart1_going_back_after_applying_part.webm) ? We should perhaps re-initialize the partitioning in the backend on "Back". Also add test for this.
- [ ] [INSTALLER-3449](https://issues.redhat.com/browse/INSTALLER-3449) Improve visibility of the partitioning choice (something like blue outline...).
- [x] Require the (more or less) same pw quality as in current Anaconda.
- [ ] [INSTALLER-3448](https://issues.redhat.com/browse/INSTALLER-3448) Add rules to passphrase checking (static and/or dynamic based on output of the pw checker (taken from cockpit))  
- [ ] InstallationDestination file and component could be renamed (commented in the PR)
- [ ] Audit error handling of promise calls.
- [ ] Move `applyConfiguration` to a better place.
- [ ] Migrate `applyConfiguration` to `async` syntax.
- [ ] Use total disk space check already on `Storage Devices` screen.
- [ ] Unify `HelpDrawer` and `DetailDrawer` component.
- [ ] Add missing aria properties.
- [ ] Consider encapsulating wizard steps into object with methods.

 
Notes:
- An issue independent of this PR [INSTALLER-3029](https://issues.redhat.com/browse/INSTALLER-3029) (Keeping the disk selection when going back) is fixed in `webui: Keep selection... ` of this set, also a test has been added (`webui: test that disk selection...`)

rvykydal commented:
The mockup:
https://www.sketch.com/s/8a7bb4fe-0797-4eed-84ef-8c507396f8c3

Screenshots:
https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screenshots/

Screencasts:
Simple walkthrough with encryption: https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screencasts/pr_autopart1_simple_encrypt.webm
Partitioning options based on selected disks: https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screencasts/pr_autopart1_partitioning_choice.webm

M4rtinK commented:
I've created a custom boot iso with this PR (state as of today): 
https://fedorapeople.org/groups/anaconda/webui/temp/f39_webui_boot_pr4599_v1.iso

rvykydal commented:
> @rvykydal ,
> 
>     * There doesn't seem to be spacing between the radio button selection label and the warning message. I don't know if the PF plain inline alert includes spacing, or if it needs a spacer added (16, 24...) https://www.patternfly.org/v4/guidelines/spacers. It looks like there is ample space to the right.
> 

Yes, I have it in my TODO list in the first comment. (In the demo I had used a bit wider viewport where the spaceItems I defined in a wrapping Flex seemed to work). Thank you for the hint with the spacer!

>     * The radio button text labels ("Erase devices and install", "Learn more" button) don't appear to be vertically centered in the data row (even when there is no error). Is there something shifting them off center, other than the wrapping message?

I couldn't figure it out quick enough, the added space at the bottom of a choice seems to be a part of DataListItem which I use for a choice in place of DataListRow (which does not add the space) because the item has the vertical lines. Definitely I need to play with this more. It is missing in my TODO list, i'll add it, thank you!

rvykydal commented:
> > @rvykydal ,
> > ```
> > * There doesn't seem to be spacing between the radio button selection label and the warning message. I don't know if the PF plain inline alert includes spacing, or if it needs a spacer added (16, 24...) https://www.patternfly.org/v4/guidelines/spacers. It looks like there is ample space to the right.
> > ```
> 
> Yes, I have it in my TODO list in the first comment. (In the demo I had used a bit wider viewport where the spaceItems I defined in a wrapping Flex seemed to work). Thank you for the hint with the spacer!

Ah, actually I am already using a spacer in a Flex element. It just stops working in this particular (+/- 20 pixels) viewport witdth. I played with it a bit more, but was not able to figure it out. Perhaps PF does not use the spacer in some limit situations, maybe it is also because of the specific outer or inner elements I am using... needs more investigating - I'm now more inclined to leave it as a separate task out of this PR, so that more experienced / lucky people can resolve it.

lclay2 commented:
@rvykydal , I don't see the spinner screen, displayed when the storage is being checked. Is it still there?

rvykydal commented:
> Hi, I noticed that the "check disk" spinner screen is presented after the Disk encryption step, rather than after selecting Storage device(s). Is this a necessity of how it works on the backend? 

Yes, we are checking (calling applyPartitioning on the backend) also the selected partitioning scheme, and technically (if I am not mistaken) it can be (reasonably) done only together with the encryption request and password provided to the backend. So the wording should be "Checking Storage Configuration" rather than "Checking Disks".

> I ask because if the storage check fails, then would the user would need to go back a couple of steps to select another device? If so, I have a concern that that could be arduous for the user, particularly if they went through the whole password creation exercise.

I think that given the content of the screens (especially the password, but also partitioning and encryption choices) is kept when going Back and Next again I didn't feel it as a big issue.

lclay2 commented:
@rvykydal , I'm adding a couple of items for the "address" list, so I can approve the PR and not hold you up:
- The mockup has a hover tip on the disabled radio button (with placeholder microcopy.) It would be nice to have this tip rather than the no-entry symbol that's in the implementation currently. It will help let the user know that the option is disabled. 
- It would be great to explore your suggestion of placing the "free space" inline alert below the disk name, maybe that will fix the alignment and radio button indenting issues that are there currently.

rvykydal commented:
@KKoukiou, Hello, Katarina, would you be able to find some time for a review? We'd need eyes of someone who knows, not learns, the stuff... Thank you.

rvykydal commented:
Updated content of partitioning options details.
I've started with what we had for the same options in RHEL 6 (https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html-single/installation_guide/index#s1-diskpartsetup-x86).
The screenshots:
https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screenshots/autopart_detail_update/
The code: see these commits of the PR:
[webui: add content for autopartitioning options details](https://github.com/rhinstaller/anaconda/pull/4599/commits/60bb1901505e2543eceb7d978a90dabf677e1683)
[webui: make not enough space warning in detail a phrase](https://github.com/rhinstaller/anaconda/pull/4599/commits/c3772027a1423df95af2229b79faa768b35226df)

lclay2 commented:
@rvykydal - I just wanted to add a note about the microcopy being placeholder until a content review has been completed, which I'm assuming is a separate issue/task. Thanks

rvykydal commented:
> @rvykydal , I'm adding a couple of items for the "address" list, so I can approve the PR and not hold you up:

@lclay2 thank you,

> 
>     * The mockup has a hover tip on the disabled radio button (with placeholder microcopy.) It would be nice to have this tip rather than the no-entry symbol that's in the implementation currently. It will help let the user know that the option is disabled.

You are right, on the preview image the tool tip does not work. Locally in Firefox it worked for me (it can be seen also on the screencast https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screencasts/pr_autopart1_partitioning_choice.webm at ~ 0:38). In any case it needs to be fixed so I'll create an item for it. Including an option to remove the no-entry icon for consideration.

> 
>     * It would be great to explore your suggestion of placing the "free space" inline alert below the disk name, maybe that will fix the alignment and radio button indenting issues that are there currently.

Good point, I'll track it in an issue as well.

M4rtinK commented:
@rvykydal fresh image with latest Rawhide & stae of the PR as of today. :)

https://fedorapeople.org/groups/anaconda/webui/temp/pr4599_2023-03-23_boot.iso

lclay2 commented:
@rvykydal , I will add things to the to-do next list, but I wanted to call out a couple of typos in case they are easy fixes. :) **Storage devices step screen**
- Select a storage devices(s)
**Disk encryption step screen**
- Encrypt the selected devices(s)?
With the (s) you have double plurals. Also I have been informed by Content, that we shouldn't be using the parens in any case.

VladimirSlavik reviewed (approved):
Looks good to me, as far as I can tell... Thank you!

lclay2 reviewed (approved):
First iteration LGTM, thanks!

KKoukiou reviewed (commented):
Need to also do an interactive review - did not checkout the branch locally yest

rvykydal commented:
I've updated the PR to handle the issues with microcopies containing devices(s) and similar strings. Also changed "Must match" to "Passphrases must match". The screenshots with new strings are here:
https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screenshots/s_strings_update/
I've also added screenshots with disabled partitioning options hints and details:
https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screenshots/disabled_autopart_options/

rvykydal commented on ui/webui/src/components/storage/DiskEncryption.jsx:
```diff
@@ -0,0 +1,314 @@
+        <AnacondaPage title={_("Encrypt the selected devices(s)?")}>
+            <TextContent>
+                {_(
+                    "You can create encrypted devices During installation." +
```
Updated:
https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screenshots/s_strings_update/updates_for_s_strings_disk_encryption.png

rvykydal commented on ui/webui/src/components/storage/InstallationDestination.jsx:
```diff
@@ -454,26 +456,8 @@ export const InstallationDestination = ({ idPrefix, setIsFormValid, onAddErrorNo
-
     return (
-        <AnacondaPage title={_("Installation destination")}>
+        <AnacondaPage title={_("Select a storage devices(s)")}>
```
Updated:
https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screenshots/s_strings_update/update_for_s_strings_installation_destination.png

rvykydal commented:
/kickstart-test --testtype smoke

KKoukiou reviewed (commented):
I noticed some spacing issues with the new screen on very small screen sizes. This can be also improved as a followup.
![Screenshot from 2023-04-03 13-03-04](https://user-images.githubusercontent.com/14921356/229491585-09e025ce-56b7-4254-983e-b346da5a0087.png)


Additionally the content of the data list rows, should be center aligned I think.

KKoukiou commented on ui/webui/src/components/storage/StorageConfiguration.jsx:
```diff
@@ -0,0 +1,350 @@
+                            <FlexItem />
+                            <FlexItem>
+                                <HelperText>
+                                    <HelperTextItem variant="warning" icon=<ExclamationTriangleIcon />>
```
it's generally more readable to put the <ExclamationTriangleIcon /> inside curly brackets {}.

rvykydal replied:
Ah, you mean `icon={<ExclamationTriangleIcon />}` I guess, I'll use the brackets.

KKoukiou commented on ui/webui/src/components/storage/StorageConfiguration.jsx:
```diff
@@ -0,0 +1,350 @@
+                    <DataListCell key="more">
+                        {scenarioAvailability[scenario.id].reason &&
+                        <Flex spaceItems={{ default: "spaceItems2xl" }}>
+                            <FlexItem />
```
I just tried to remove it and only add this cell if the element inside exists and it works as expected. Only made the first column normal cell - DataListAction is more about alignment of the last column.

``` diff
diff --git ui/webui/src/components/storage/StorageConfiguration.jsx ui/webui/src/components/storage/StorageConfiguration.jsx
index ec6566ba99..c023e1d527 100644
--- ui/webui/src/components/storage/StorageConfiguration.jsx
+++ ui/webui/src/components/storage/StorageConfiguration.jsx
@@ -267,7 +267,7 @@ const GuidedPartitioning = ({ idPrefix, scenarios, setIsFormValid }) => {
         <DataListItem key={scenario.id}>
             <DataListItemRow>
                 <DataListItemCells dataListCells={[
-                    <DataListAction key="radio">
+                    <DataListCell key="radio">
                         {!scenarioAvailability[scenario.id].available &&
                         <Tooltip
                           aria-live="polite"
@@ -283,21 +283,18 @@ const GuidedPartitioning = ({ idPrefix, scenarios, setIsFormValid }) => {
                           isChecked={selectedScenario === scenario.id}
                           onChange={() => onScenarioToggled(scenario.id)}
                         />
-                    </DataListAction>,
-                    <DataListCell key="more">
-                        {scenarioAvailability[scenario.id].reason &&
-                        <Flex spaceItems={{ default: "spaceItems2xl" }}>
-                            <FlexItem />
-                            <FlexItem>
+                    </DataListCell>,
+                    ...scenarioAvailability[scenario.id].reason
+                        ? [
+                            <DataListCell key="more">
                                 <HelperText>
                                     <HelperTextItem variant="warning" icon=<ExclamationTriangleIcon />>
                                         {scenarioAvailability[scenario.id].reason}
                                     </HelperTextItem>
                                 </HelperText>
-                            </FlexItem>
-                            <FlexItem />
-                        </Flex>}
-                    </DataListCell>,
+                            </DataListCell>
+                        ]
+                        : [],

```

M4rtinK reviewed (approved):
Looks good to me as well, thanks a lot Radek! :)

rvykydal commented:
> I noticed some spacing issues with the new screen on very small screen sizes. This can be also improved as a followup. ![Screenshot from 2023-04-03 13-03-04](https://user-images.githubusercontent.com/14921356/229491585-09e025ce-56b7-4254-983e-b346da5a0087.png)
> 
> Additionally the content of the data list rows, should be center aligned I think.

Thank you for looking it again. I think it is already tracked in the todo list on top in https://issues.redhat.com/browse/INSTALLER-3443 and https://issues.redhat.com/browse/INSTALLER-3444. I actually already started working on that so far moving the radio out of the cell list (https://github.com/rvykydal/anaconda/commit/1170de0c1d3a42aafcf59d9f7e9795d7720714a8) which helped to keep the spacing. More details are in the jira issue. I'll try your suggestion. However, @lclay2 seems to prefer another approach (described in https://issues.redhat.com/browse/INSTALLER-3443) so I'll explore that as well. Anyway, thank your suggestions!

rvykydal commented:
> > ... However, @lclay2 seems to prefer another approach (described in https://issues.redhat.com/browse/INSTALLER-3443) so I'll explore that as well. ...

Here is the alternative approach:
https://rvykydal.fedorapeople.org/webui/autopart_pr1/v1/screenshots/partitioning_spacing_update_v2/
https://github.com/rvykydal/anaconda/commit/3c5cd45bdb180f0e81a16fbf6859db1fe24c6476

lclay2 commented:
@rvykydal , thanks for the examples of the alternative approach. It looks good to me! Seems like it might be less brittle than the prior solution, wdyt?

rvykydal commented:
> @rvykydal , thanks for the examples of the alternative approach. It looks good to me! Seems like it might be less brittle than the prior solution, wdyt?

I agree, I opened a PR for the update: https://github.com/rhinstaller/anaconda/pull/4674
